src/web/metrics.rs

```diff
@@ -2,42 +2,65 @@ use crate::{
     db::Pool, metrics::duration_to_seconds, utils::spawn_blocking, web::error::AxumResult,
     BuildQueue, Config, Metrics,
 };
-use anyhow::Context as _;
+use anyhow::{Context as _, Result};
 use axum::{
-    body::Body,
     extract::{Extension, MatchedPath},
     http::Request as AxumRequest,
-    http::{
-        header::{CONTENT_LENGTH, CONTENT_TYPE},
-        Response as AxumHttpResponse, StatusCode,
-    },
+    http::{header::CONTENT_TYPE, StatusCode},
     middleware::Next,
     response::IntoResponse,
 };
-use prometheus::{Encoder, HistogramVec, TextEncoder};
+use prometheus::{proto::MetricFamily, Encoder, HistogramVec, TextEncoder};
 use std::{borrow::Cow, sync::Arc, time::Instant};
 #[cfg(test)]
 use tracing::debug;
 
+async fn fetch_and_render_metrics(
+    fetcher: impl Fn() -> Result<Vec<MetricFamily>> + Send + 'static,
+) -> AxumResult<impl IntoResponse> {
+    let metrics_families = spawn_blocking(fetcher).await?;
+
+    let mut buffer = Vec::new();
+    TextEncoder::new()
+        .encode(&metrics_families, &mut buffer)
+        .context("error encoding metrics")?;
```

Comment:
I would move the encoding inside the spawned task too, no need to block the async threads on it if we're already doing cross-task stuff.

Reply:
valid, done